aspnetcore/blazor/security/webassembly/standalone-with-identity/qrcodes-for-authenticator-apps.md

```diff
@@ -0,0 +1,853 @@
+    string password, 
+    string twoFactorCode);
+
+public Task<FormResult> LoginTwoFactorRecoveryCodeAsync(
+    string email, 
+    string password, 
+    string twoFactorRecoveryCode);
+
+public Task<TwoFactorResponse> TwoFactorRequestAsync(
+    TwoFactorRequest twoFactorRequest);
+```
+
+## Update the cookie authentication state provider
+
+Update the `CookieAuthenticationStateProvider` with features to add the following features:
+
+* Authenticate users with either a TOTP authenticator app code or a recovery code.
+* Manage 2FA in the app.
+
+At the top of the `CookieAuthenticationStateProvider.cs` file, add a `using` statement for <xref:System.Text.Json.Serialization?displayProperty=fullName>:
+
+```csharp
+using System.Text.Json.Serialization;
+```
+
+In the <xref:System.Text.Json.JsonSerializerOptions>, add the <xref:System.Text.Json.JsonSerializerOptions.DefaultIgnoreCondition> option set to <xref:System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull?displayProperty=nameWithType>, which avoids serializing null properties:
+
+```diff
+private readonly JsonSerializerOptions jsonSerializerOptions =
+    new()
+    {
+        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
++       DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
+    };
+```
+
+The `LoginAsync` method is updated with the following logic:
+
+* Attempt a normal login at the `/login` endpoint with an email address and password.
+* If the server responds with a success status code, the method returns a `FormResult` with the `Succeeded` property set to `true`.
+* If the server responds with the *401 - Unauthorized* status code and a detail code of "`RequiresTwoFactor`," a `FormResult` is returned with `Succeeded` set to `false` and the `RequiresTwoFactor` detail in the error list.
+
+In `Identity/CookieAuthenticationStateProvider.cs`, replace the `LoginAsync` method with the following code:
+
+```csharp
+public async Task<FormResult> LoginAsync(string email, string password)
+{
+    try
+    {
+        var result = await httpClient.PostAsJsonAsync(
+            "login?useCookies=true", new
+            {
+                email,
+                password
+            });
+
+        if (result.IsSuccessStatusCode)
+        {
+            NotifyAuthenticationStateChanged(GetAuthenticationStateAsync());
+
+            return new FormResult { Succeeded = true };
+        }
+        else if (result.StatusCode == HttpStatusCode.Unauthorized)
+        {
+            var responseJson = await result.Content.ReadAsStringAsync();
+            var response = JsonSerializer.Deserialize<LoginResponse>(
+                responseJson, jsonSerializerOptions);
+
+            if (response?.Detail == "RequiresTwoFactor")
+            {
+                return new FormResult
+                {
+                    Succeeded = false,
+                    ErrorList = [ "RequiresTwoFactor" ]
+                };
+            }
+        }
+    }
+    catch { }
```

Comment:
But this is mirroring your approach ... from the code that you submitted ...

https://github.com/dotnet/blazor-samples/blob/main/9.0/BlazorWebAssemblyStandaloneWithIdentity/BlazorWasmAuth/Identity/CookieAuthenticationStateProvider.cs#L98

If we remove the try-catch, then how can it return a `FormResult` on an error with `Succeeded` of `false` and the error message for the user?

Reply:
That question AND one more note ...

I had to add ...

```csharp
await Acct.TwoFactorRequestAsync(new() { ForgetMachine = true });
```

... to the `Disable2fa` method because if it isn't there and 2FA is disabled by the user and re-enabled, it remembers them on the next login attempt and doesn't ask for a 2FA code. By forgetting the machine when 2FA is disabled, it does request a new 2FA/recovery code on the next login after they've re-enabled 2FA. I think that's a better UE.

One more problem has cropped up. If the user disables 2FA and then re-enables it, I would kind'a expect a new set of recovery codes to get generated. As it stands, the user would have to click the button to get them. I'm going to make it so that a new set of recovery codes is always generated after 2FA is enabled ... I'll do that by checking to see if 2FA enable was successful, and then if there are no recovery codes, I'll make it generate a new batch. Apparently, if they're still on the user's account from the last time that they had them established, the API doesn't generate a new set. I feel that a better UE is if they disable and re-enable that they get a new set of codes. I'll take care of that in the morning. I'm off to 🛌💤 shortly.